c++/src/kj/async.h

```diff
@@ -585,6 +585,114 @@ inline CaptureByMove<Func, Decay<MovedParam>> mvCapture(MovedParam&& param, Func
+#if KJ_HAS_COROUTINE
+
+namespace _ {
+
+template<typename Functor>
+struct CaptureForCoroutine {
+  kj::Maybe<Functor> maybeFunctor;
+
+  explicit CaptureForCoroutine(Functor&& f) : maybeFunctor(kj::mv(f)) {}
+
+  template<typename ...Args>
+  static auto coInvoke(Functor functor, Args&&... args)
+      -> decltype(functor(kj::fwd<Args>(args)...)) {
+    // Since the functor is now in the local scope and no longer a member variable, it will be
+    // persisted in the coroutine state.
+
+    // Note that `co_await functor(...)` can still return `void`. It just happens that
+    // `co_return voidReturn();` is explicitly allowed.
+    co_return co_await functor(kj::fwd<Args>(args)...);
+  }
+
+  template<typename ...Args>
+  auto operator()(Args&&... args) {
+    KJ_IREQUIRE(maybeFunctor != nullptr,
+        "Attempted to invoke CaptureForCoroutine functor multiple times");
+    auto localFunctor = kj::mv(*kj::_::readMaybe(maybeFunctor));
+    maybeFunctor = nullptr;
+    return coInvoke(kj::mv(localFunctor), kj::fwd<Args>(args)...);
+  }
+};
+
+}  // namespace _
+
+template <typename Functor>
+auto coCapture(Functor&& f) {
+  // Assuming `f()` returns a Promise<T> `p`, wrap `f` in such a way that it will outlive its
+  // returned Promise. Note that the returned object may only be invoked once.
+  //
+  // This function is meant to help address this pain point with functors that return a coroutine:
+  // https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rcoro-capture
+  //
+  // The two most common patterns where this may be useful look like so:
+  // ```
+  // void addTask(Value myValue) {
+  //   auto myFun = [myValue]() -> kj::Promise<void> {
+  //     ...
+  //     co_return;
+  //   };
+  //   tasks.add(myFun());
+  // }
+  // ```
+  // and
+  // ```
+  // kj::Promise<void> afterPromise(kj::Promise<void> promise, Value myValue) {
+  //   auto myFun = [myValue]() -> kj::Promise<void> {
+  //     ...
+  //     co_return;
+  //   };
+  //   return promise.then(kj::mv(myFun));
+  // }
+  // ```
+  //
+  // Note that there are potentially more optimal alternatives to both of these patterns:
+  // ```
+  // void addTask(Value myValue) {
+  //   auto myFun = [](auto myValue) -> kj::Promise<void> {
+  //     ...
+  //     co_return;
+  //   };
+  //   tasks.add(myFun(myValue));
+  // }
+  // ```
+  // and
+  // ```
+  // kj::Promise<void> afterPromise(kj::Promise<void> promise, Value myValue) {
+  //   auto myFun = [&]() -> kj::Promise<void> {
+  //     ...
+  //     co_return;
+  //   };
+  //   co_await promise;
+  //   co_await myFun();
+  //   co_return;
+  // }
+  // ```
+  //
+  // For situations where you are trying to capture a specific local variable, kj::mvCapture() can
+  // also be useful:
+  // ```
+  // kj::Promise<void> reactToPromise(kj::Promise<MyType> promise) {
+  //   BigA a;
+  //   TinyB b;
+  //
+  //   doSomething(a, b);
+  //   return promise.then(kj::mvCapture(b, [](TinyB b, MyType type) -> kj::Promise<void> {
```

Comment:
Out of curiosity, is there a need for `kj::mvCapture` (which got deprecated [recently](https://github.com/capnproto/capnproto/pull/1558)) instead of using a C++14 generalized lambda?

```
return promise.then([b=kj::mv(b)](MyTypetype) -> kj::Promise<void> {
```

Reply:
Yes, this whole comment block is discussing a problem with C++20 coroutines (`co_await`/`co_return`) when combined with lambads -- things captured in the `[]` capture block typically no longer exist after the first `co_await`, but can still be referenced, leading to undefined behavior. Parameters, however, do not suffer this problem, since they are copied into the coroutine frame. So it turns out `kj::mvCapture` is, unintentionally, a possible work-around for this problem, since it uses parameters instead of `[]`. But `coCapture` is probably a better solution.